src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp

```diff
@@ -0,0 +1,982 @@
+ *          \alpha_1\left(c^{l+1,m,n}_{0,0,0}-c^{l,m,n}_{0,0,0}\right),
+ *          \alpha_1\left(c^{l,m,n}_{0,0,0}-c^{l-1,m,n}_{0,0,0}\right)
+ *     \right),\\
+ * \tilde{c}^{l,m,n}_{0,1,0}=\mathrm{minmod}
+ *   &\left(c_{0,1,0}^{l,m,n},
+ *          \alpha_1\left(c^{l,m+1,n}_{0,0,0}-c^{l,m,n}_{0,0,0}\right),
+ *          \alpha_1\left(c^{l,m,n}_{0,0,0}-c^{l,m-1,n}_{0,0,0}\right)
+ *     \right),\\
+ * \tilde{c}^{l,m,n}_{0,0,1}=\mathrm{minmod}
+ *   &\left(c_{0,0,1}^{l,m,n},
+ *          \alpha_1\left(c^{l,m,n+1}_{0,0,0}-c^{l,m,n}_{0,0,0}\right),
+ *          \alpha_1\left(c^{l,m,n}_{0,0,0}-c^{l,m,n-1}_{0,0,0}\right)
+ *     \right),
+ * \f}
+ *
+ *
+ * Finally, limiting is performed as follows:
+ *
+ * - limit \f$c_{2,2,2}\f$, if not changed, stop
+ * - limit \f$c_{2,2,1}\f$, \f$c_{2,1,2}\f$, and \f$c_{1,2,2}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{2,2,0}\f$, \f$c_{2,0,2}\f$, and \f$c_{0,2,2}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{2,1,1}\f$, \f$c_{1,2,1}\f$, and \f$c_{1,1,2}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{2,1,0}\f$, \f$c_{2,0,1}\f$, \f$c_{1,2,0}\f$, \f$c_{1,0,2}\f$,
+ * \f$c_{0,1,2}\f$, and \f$c_{0,2,1}\f$, if all not changed, stop
+ * - limit \f$c_{2,0,0}\f$, \f$c_{0,2,0}\f$, and \f$c_{0,0,2}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{1,1,1}\f$, if not changed, stop
+ * - limit \f$c_{1,1,0}\f$, \f$c_{1,0,1}\f$, and \f$c_{0,1,1}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{1,0,0}\f$, \f$c_{0,1,0}\f$, and \f$c_{0,0,1}\f$, if all not
+ * changed, stop
+ *
+ * The 1d and 2d implementations are straightforward restrictions of the
+ * described algorithm.
+ *
+ * #### Limitations:
+ *
+ * - We are recomputing the spectral coefficients after sending the coefficients
+ *   to the neighbor. We should be able to avoid this either by storing the
+ *   coefficients in the DataBox or by allowing the limiters' `packaged_data`
+ *   function to return an object to be passed as an additional argument to the
+ *   `operator()` (still would need to be stored in the DataBox).
+ * - We cannot handle the case where neighbors have more/fewer
+ *   coefficients. In the case that the neighbor has more coefficients we can
+ *   just ignore the higher coefficients. In the case that the neighbor has
+ *   fewer coefficients we have a few choices that will probably depend on which
+ *   specific HL is used.
+ * - Having different number of collocation points in different directions is
+ *   not supported. However, it is straightforward to handle this case. The
+ *   outer most loop is in the direction with the most collocation points, while
+ *   the inner most loop is over the direction with the fewest collocation
+ *   points. The highest to lowest coefficients can then be limited
+ *   appropriately again.
+ * - h-refinement is not supported, but is straightforward to add since the
+ *   differences in coefficients in Eq.\f$\ref{eq:krivodonova 3d minmod}\f$ will
+ *   just be multiplied by non-unity factors.
```

Comment:
It seems like you comment on a number of things here that are straightforward to add. Which will we actually need in the near future? For the shock tests, support for p-anisotropic meshes would be useful. With AMR coming up, h-refinement will be important. Are you planning to add these features in upcoming PRs?

Reply:
All this depends on how well the limiter works. I don't see why we need anisotropic p for the shock tests. I'd argue that's actually the wrong thing to test.

Reply:
I am planning to add these in the future, but how soon will depend on how well the limiter works on the disk. If it doesn't perform well then it'll be a while out, if it performs well it'll be sooner.

Reply:
👍 on the extended mesh checks

Reply:
It is not at all obvious to me that the h-refinement case is straightfoward. With multiple neighbors across a face, it is not obvious how the relationships between the spectral coeffs of neighboring cells are affected. If you have it all worked out, that's fine, but otherwise I wouldn't claim too much here.